lib/CL/pocl_llvm_api.cc

```diff
@@ -1527,10 +1561,37 @@ static PassManager& kernel_compiler_passes
   passes.push_back("domtree");
   if (device->autolocals_to_args)
 	  passes.push_back("automatic-locals");
+
   if (SPMDDevice)
     passes.push_back("flatten-inline-all");
-  else
-    passes.push_back("flatten-globals");
+  else {
+    /* LLVM is either still not intelligent enough to flatten,
+     * or we're missing some pass.
+     *
+     * The only reason we've reintroduced "global-only" flatten was that
```

Comment:
What do you mean global-only here? Could we simply rely on LLVM inliner nowadays? That is, flatten (this used to mean "inline everything") only those that have to be (functions which take in the context struct basically), but let the rest of the inlining decisions to LLVM. This should work fine for CPUs and DSPs. For GPUs we might want to inline more aggressively.